packages/@aws-cdk-testing/cli-integ/lib/staging/maven.ts

```diff
@@ -0,0 +1,87 @@
+/* eslint-disable no-console */
+import * as path from 'path';
+import { writeFile } from '../files';
+import { shell } from '../shell';
+import { LoginInformation } from './codeartifact';
+import { parallelShell } from './parallel-shell';
+import { UsageDir } from './usage-dir';
+
+// Do not try to JIT the Maven binary
+const NO_JIT = '-XX:+TieredCompilation -XX:TieredStopAtLevel=1';
+
+export async function mavenLogin(login: LoginInformation, usageDir: UsageDir) {
+  await writeMavenSettingsFile(settingsFile(usageDir), login);
+
+  // Write env var
+  // Twiddle JVM settings a bit to make Maven survive running on a CodeBuild box.
+  await usageDir.addToEnv({
+    MAVEN_OPTS: `-Duser.home=${usageDir.directory} ${NO_JIT} ${process.env.MAVEN_OPTS ?? ''}`.trim(),
+  });
+}
+
+function settingsFile(usageDir: UsageDir) {
+  // If we configure usageDir as a fake home directory Maven will find this file.
+  // (No other way to configure the settings file as part of the environment).
+  return path.join(usageDir.directory, '.m2', 'settings.xml');
+}
+
+export async function uploadJavaPackages(packages: string[], login: LoginInformation, usageDir: UsageDir) {
+  await parallelShell(packages, async (pkg, output) => {
+    console.log(`⏳ ${pkg}`);
+
+    await shell(['mvn',
+      `--settings=${settingsFile(usageDir)}`,
+      'org.apache.maven.plugins:maven-deploy-plugin:3.0.0:deploy-file',
+      `-Durl=${login.mavenEndpoint}`,
+      '-DrepositoryId=codeartifact',
+      `-DpomFile=${pkg}`,
+      `-Dfile=${pkg.replace(/.pom$/, '.jar')}`,
+      `-Dsources=${pkg.replace(/.pom$/, '-sources.jar')}`,
+      `-Djavadoc=${pkg.replace(/.pom$/, '-javadoc.jar')}`], {
+      output,
+      modEnv: {
+        // Do not try to JIT the Maven binary
+        MAVEN_OPTS: `${NO_JIT} ${process.env.MAVEN_OPTS ?? ''}`.trim(),
+      },
+    });
+
+    console.log(`✅ ${pkg}`);
+  },
+  (pkg, output) => {
+    if (output.toString().includes('409 Conflict')) {
+      console.log(`❌ ${pkg}: already exists. Skipped.`);
+      return 'skip';
+    }
+    return 'fail';
+  });
+}
+
+export async function writeMavenSettingsFile(filename: string, login: LoginInformation) {
```

Comment:
[high number of file changes, I might just have scrolled past] The "filename" (`usageDir`) should be removed in a `cleanup()`, right/ does it?

Reply:
`usageDir` is (may be) cleaned up in a separate invocation. The CLI also has two possible APIs:

* `publish` + `clean`; or
* `run <cmd>`